### PR TITLE
3d map icons

### PR DESCRIPTION
**Pull request by nnngrach**
https://github.com/osmandapp/OsmAnd-iOS/issues/3778

turn on online plugin
Screenshot 2024-07-04 at 12 32 55

select new appeared 3d icons
Screenshot 2024-07-04 at 12 36 11

---

2d
Screenshot 2024-07-04 at 12 33 33
3d
Screenshot 2024-07-04 at 12 34 14

---

2d
Screenshot 2024-07-04 at 12 35 23
3d
Screenshot 2024-07-04 at 12 35 41

---

Folder structure:

Screenshot 2024-07-04 at 12 45 09 Screenshot 2024-07-04 at 12 44 48

---

TODO: There is a bug with layers order. 2d layer with heading direction is showing above 3d layer with icon. There' is android's fix:
https://github.com/osmandapp/OsmAnd/pull/20184/files

(Screenshot with deleted heading layer. Just for example)
Screenshot 2024-07-01 at 21 58 37

### CODE REVIEW

File: Sources/Controllers/Map/Layers/OAMyPositionLayer.mm [hunk 1]
```diff
@@ -469,83 +478,203 @@ - (void) generateMarkersCollection
...
-        locationAndCourseMarkerBuilder.addOnMapSurfaceIcon(c.locationMainIconKeyLostNight,
-                                                           OsmAnd::SingleSkImage(locationMainLostNightIcon));
-        c.locationMarkerLostNight = locationAndCourseMarkerBuilder.buildAndAddToCollection(c.markerCollection);
-    
-        [self updateMode:c];
+        OANavigationIcon *navIcon = [OANavigationIcon withIconName:navigationIconName];
+        if ([navIcon isModel])
+        {
+            navigationModel = [Model3dHelper.shared getModelWithModelName:navigationIconName callbackOnLoad:^(OAModel3dWrapper * _Nullable model) {
+                [self refreshMarkersCollection];
```
Comment (tigrim) [inline Sources/Controllers/Map/Layers/OAMyPositionLayer.mm:503]: memory leak

File: Sources/Controllers/Map/Layers/OAMyPositionLayer.mm [hunk 2]
```diff
@@ -469,83 +478,203 @@ - (void) generateMarkersCollection
...
+            }
+        }
+        
+        OALocationIcon *locIcon = [OALocationIcon withIconName:locationIconName];
+        if ([locIcon isModel])
+        {
+            if (![navIcon isModel] || navigationModel)
+            {
+                locationModel = [Model3dHelper.shared getModelWithModelName:locationIconName callbackOnLoad:^(OAModel3dWrapper *_Nullable model) {
+                    [self refreshMarkersCollection];
```
Comment (tigrim) [inline Sources/Controllers/Map/Layers/OAMyPositionLayer.mm:519]: up

File: Sources/Controllers/Map/Layers/OAMyPositionLayer.mm [hunk 3]
```diff
@@ -469,83 +478,203 @@ - (void) generateMarkersCollection
...
+        if (locationModel)
+        {
+            [locationModel setMainColor:iconColor];
+            locationModelCpp = [locationModel model];
+            locationHeadingSkImage = [OANativeUtilities skImageFromCGImage:[locIcon headingIconWithColor:iconColor].CGImage];
+            
+            c.locationMainIconKeyDay = reinterpret_cast<OsmAnd::MapMarker::OnSurfaceIconKey>(1);
+            locationAndCourseMarkerBuilder.setModel3D(locationModelCpp);
+            
+
```
Comment (tigrim) [inline Sources/Controllers/Map/Layers/OAMyPositionLayer.mm:602]: remove /n

File: Sources/Controllers/Settings/OAProfileAppearanceViewController.mm [hunk 4]
```diff
@@ -357,16 +368,58 @@ - (void) generateData
                @"ic_action_light_aircraft"];
 }
 
-- (NSArray<UIImage *> *) getIconsAtRest
+- (NSArray<NSString *> *) getlocationIconNames
 {
+    NSMutableArray<NSString *> *iconNames = [NSMutableArray array];
```
Comment (tigrim) [inline Sources/Controllers/Settings/OAProfileAppearanceViewController.mm:373]:
```
NSMutableArray<NSString *> *iconNames = [@[
    LOCATION_ICON_DEFAULT,
    LOCATION_ICON_CAR,
    LOCATION_ICON_BICYCLE
] mutableCopy];
```

File: Sources/Controllers/Settings/OAProfileAppearanceViewController.mm [hunk 5]
```diff
@@ -357,16 +368,58 @@ - (void) generateData
...
+    [iconNames addObject:LOCATION_ICON_CAR];
+    [iconNames addObject:LOCATION_ICON_BICYCLE];
+    
+    Model3dHelper *helper = [Model3dHelper shared];
+    NSArray<NSString *> *model3dIconsNames = [Model3dHelper listModels];
+    for (NSString *modelIconName in model3dIconsNames)
+    {
+        [iconNames addObject:modelIconName];
+    }
+    return iconNames;
```
Comment (tigrim) [inline Sources/Controllers/Settings/OAProfileAppearanceViewController.mm:384]: Here we don't need 'for'. Just add the array.
    
```
[iconNames addObjectsFromArray:model3dIconsNames];
    return [iconNames copy]; 
```

File: Sources/Controllers/Settings/OAProfileAppearanceViewController.mm [hunk 6]
```diff
@@ -357,16 +368,58 @@ - (void) generateData
...
+    [iconNames addObject:NAVIGATION_ICON_NAUTICAL];
+    [iconNames addObject:NAVIGATION_ICON_CAR];
+    
+    Model3dHelper *helper = [Model3dHelper shared];
+    NSArray<NSString *> *model3dIconsNames = [Model3dHelper listModels];
+    for (NSString *modelIconName in model3dIconsNames)
+    {
+        [iconNames addObject:modelIconName];
+    }
+    return iconNames;
```
Comment (tigrim) [inline Sources/Controllers/Settings/OAProfileAppearanceViewController.mm:411]: up

File: Sources/Helpers/Model3dHelper/Model3dHelper.swift [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+    private var failedModels = Set<String>()
+    
+    private override init() {
+        app = OsmAndApp.swiftInstance()
+        settings = OAAppSettings.sharedManager()
+    }
+    
+    func getModel(modelName: String, callbackOnLoad: callbackWithModel3d?) -> OAModel3dWrapper? {
+        if !modelName.hasPrefix(MODEL_NAME_PREFIX) {
+            if let callbackOnLoad {
```
Comment (tigrim) [inline Sources/Helpers/Model3dHelper/Model3dHelper.swift:32]: guard

File: Sources/Helpers/Model3dHelper/Model3dHelper.swift [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+        let modelDirPath = app.documentsPath.appendingPathComponent(MODEL_3D_DIR).appendingPathComponent(modelName)
+        if !Model3dHelper.isModelExist(dir: modelDirPath) {
+            return
+        }
+        
+        modelsInProgress.insert(modelName)
+        
+        let task = OALoad3dModelTask(modelName) { model in
+            if model == nil {
+                self.failedModels.insert(modelName)
```
Comment (tigrim) [inline Sources/Helpers/Model3dHelper/Model3dHelper.swift:71]: retain cycle

File: Sources/Helpers/Model3dHelper/Model3dHelper.swift [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+            let modelsDirs = try FileManager.default.contentsOfDirectory(atPath: modelsDir)
+            if !modelsDirs.isEmpty {
+                for model in modelsDirs {
+                    if isModelExist(dir: modelsDir.appendingPathComponent(model)) {
+                        modelsDirNames.append(MODEL_NAME_PREFIX + model)
+                    }
+                }
+            }
+        } catch {
+        }
```
Comment (tigrim) [inline Sources/Helpers/Model3dHelper/Model3dHelper.swift:97]: add DebugPrint and message error

File: Sources/Helpers/Model3dHelper/Model3dHelper.swift [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+            let modelFiles = try FileManager.default.contentsOfDirectory(atPath: dir)
+            if !modelFiles.isEmpty {
+                for file in modelFiles {
+                    if file.hasSuffix(OBJ_FILE_EXT) {
+                        return true
+                    }
+                }
+            }
+        } catch {
+        }
```
Comment (tigrim) [inline Sources/Helpers/Model3dHelper/Model3dHelper.swift:116]: up

**Comment by nnngrach** [comment]
video:
https://github.com/osmandapp/OsmAnd-iOS/assets/35684515/a51a9fe1-94d3-42cc-9fcd-189a3dc5423b

File: Sources/Controllers/Map/Layers/OAMyPositionLayer.mm [hunk 11]
```diff
@@ -469,83 +534,194 @@ - (void) generateMarkersCollection
...
+        else
+        {
+            sk_sp<SkImage> locationMainLostDayIcon = [OANativeUtilities skImageFromCGImage:[navIcon getMapIcon:UIColorFromRGB(location_icon_color_lost)].CGImage];
+            locationAndCourseMarkerBuilder.addOnMapSurfaceIcon(c.locationMainIconKeyLostDay,
+                                                               OsmAnd::SingleSkImage(locationMainLostDayIcon));
+        }
         c.locationMarkerLostDay = locationAndCourseMarkerBuilder.buildAndAddToCollection(c.markerCollection);
         
         // Lost (night)
         locationAndCourseMarkerBuilder.clearOnMapSurfaceIcons();
```
Comment (tigrim) [inline Sources/Controllers/Map/Layers/OAMyPositionLayer.mm:710]: duplicate (695) One call is not enough?

File: Sources/Controllers/Settings/OAProfileAppearanceViewController.mm [hunk 12]
```diff
@@ -357,16 +368,53 @@ - (void) generateData
...
-- (NSArray<UIImage *> *) getIconsAtRest
+- (NSArray<NSString *> *) getlocationIconNames
 {
+    NSMutableArray<NSString *> *iconNames = [@[
+        LOCATION_ICON_DEFAULT,
+        LOCATION_ICON_CAR,
+        LOCATION_ICON_BICYCLE
+    ] mutableCopy];
+    
+    Model3dHelper *helper = [Model3dHelper shared];
```
Comment (tigrim) [inline Sources/Controllers/Settings/OAProfileAppearanceViewController.mm:379]: What is this variable for?

File: Sources/Controllers/Settings/OAProfileAppearanceViewController.mm [hunk 13]
```diff
@@ -357,16 +368,53 @@ - (void) generateData
...
 
-- (NSArray<UIImage *> *) getNavIcons
+- (NSArray<NSString *> *) getNavigationIconNames
 {
+    NSMutableArray<NSString *> *iconNames = [NSMutableArray array];
+    [iconNames addObject:NAVIGATION_ICON_DEFAULT];
+    [iconNames addObject:NAVIGATION_ICON_NAUTICAL];
+    [iconNames addObject:NAVIGATION_ICON_CAR];
+    
+    Model3dHelper *helper = [Model3dHelper shared];
```
Comment (tigrim) [inline Sources/Controllers/Settings/OAProfileAppearanceViewController.mm:403]: up

File: Sources/Profiles/OANavigationIcon.m [hunk 14]
```diff
@@ -9,72 +9,124 @@
...
-+ (instancetype) withNavigationIcon:(EOANavigationIcon)navigationIcon
++ (instancetype) withIconName:(NSString *)iconName
 {
     OANavigationIcon *obj = [[OANavigationIcon alloc] init];
     if (obj)
-        obj.navigationIcon = navigationIcon;
+    {
+        NSString *migratedOldValue = [self getMigratedValue:iconName];
+        if (migratedOldValue)
+            obj.iconName = migratedOldValue;
```
Comment (tigrim) [inline Sources/Profiles/OANavigationIcon.m:30]: use ternary operator

File: Sources/Controllers/Map/Layers/OAMyPositionLayer.mm (hunk 11 above)
Comment (nnngrach) [reply Sources/Controllers/Map/Layers/OAMyPositionLayer.mm:710]: Here is a screenshot from original code from master branch. Clearing called before  of creating every new icon.

Screenshot 2024-07-05 at 16 03 37

File: Sources/Profiles/OALocationIcon.m [hunk 15]
```diff
@@ -26,10 +26,7 @@ + (instancetype) withIconName:(NSString *)iconName
     if (obj)
     {
         NSString *migratedOldValue = [self getMigratedValue:iconName];
-        if (migratedOldValue)
-            obj.iconName = migratedOldValue;
-        else
-            obj.iconName = iconName;
+        obj.iconName = migratedOldValue ? migratedOldValue : iconName;
```
Comment (tigrim) [inline Sources/Profiles/OALocationIcon.m:29]: obj.iconName = migratedOldValue ?: iconName;

File: Sources/Profiles/OANavigationIcon.m [hunk 16]
```diff
@@ -26,10 +26,7 @@ + (instancetype) withIconName:(NSString *)iconName
     if (obj)
     {
         NSString *migratedOldValue = [self getMigratedValue:iconName];
-        if (migratedOldValue)
-            obj.iconName = migratedOldValue;
-        else
-            obj.iconName = iconName;
+        obj.iconName = migratedOldValue ? migratedOldValue : iconName;
```
Comment (tigrim) [inline Sources/Profiles/OANavigationIcon.m:29]: obj.iconName = migratedOldValue ?: iconName;

**Review by tigrim — APPROVED** [review]

File: Sources/Controllers/Map/Layers/OAMyPositionLayer.mm [hunk 17]
```diff
@@ -469,83 +534,194 @@ - (void) generateMarkersCollection
...
+        std::shared_ptr<const OsmAnd::Model3D> navigationModelCpp;
+        std::shared_ptr<const OsmAnd::Model3D> locationModelCpp;
+        
+        BOOL notInited = NO;
+        
+        OANavigationIcon *navIcon = [OANavigationIcon withIconName:navigationIconName];
+        if ([navIcon isModel])
+        {
+            navigationModel = [Model3dHelper.shared getModelWithModelName:navigationIconName callbackOnLoad:^(OAModel3dWrapper * _Nullable model) {
+                [weakSelf refreshMarkersCollection];
```
Comment (alex-osm) [inline Sources/Controllers/Map/Layers/OAMyPositionLayer.mm:559]: Refresh all markers again on every model load is bad thing. You may load all models right away, but generateMarkersCollection should not be called on every load.

File: Sources/Controllers/Map/Layers/OAMyPositionLayer.mm [hunk 18]
```diff
@@ -625,11 +801,13 @@ - (void) updateMyLocationCourseProvider
     [self.mapViewController runWithRenderSync:^{
     
         OAApplicationMode *currentMode = [OAAppSettings sharedManager].applicationMode.get;
+        
         for (OAApplicationMode *mode in _modeMarkers.keyEnumerator)
         {
             OAMarkerCollection *c = [_modeMarkers objectForKey:mode];
             if (mode == currentMode)
             {
+                [c updateState];
```
Comment (alex-osm) [inline Sources/Controllers/Map/Layers/OAMyPositionLayer.mm:810]: Why is it needed?

File: Sources/Helpers/Model3dHelper/Model3dHelper.swift [hunk 19]
```diff
@@ -0,0 +1,119 @@
...
+        
+        let modelDirPath = app.documentsPath.appendingPathComponent(MODEL_3D_DIR).appendingPathComponent(modelName)
+        if !Model3dHelper.isModelExist(dir: modelDirPath) {
+            return
+        }
+        
+        modelsInProgress.insert(modelName)
+        
+        let task = OALoad3dModelTask(modelName) { [weak self] model in
```
[the remark this answers is not in the document]
Comment (nnngrach) [reply Sources/Helpers/Model3dHelper/Model3dHelper.swift:101]: 1 - moved inside.
2 - now calling only from main thread ()
before / after
Screenshot 2024-07-09 at 17 54 57 !Screenshot 2024-07-09 at 17 55 22

File: Sources/Helpers/Model3dHelper/OAModel3dHelper.mm [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+
+#import "OAModel3dHelper.h"
+#import "OAModel3dHelper+cpp.h"
+#import "OANativeUtilities.h"
+#import "OsmAndApp.h"
+#import "OAIndexConstants.h"
+
+@implementation OAModel3dWrapper
+{
+    OsmAnd::Model3D *_model;
```
[the remark this answers is not in the document]
Comment (nnngrach) [reply Sources/Helpers/Model3dHelper/OAModel3dHelper.mm:17]: I have a question how to change color for this way? 
how to change color

File: Sources/Helpers/OAAppSettings.m [hunk 21]
```diff
@@ -3993,17 +3993,17 @@ - (instancetype) init
...
         [_profilePreferences setObject:_navigationIcon forKey:@"navigation_icon"];
 
-        _locationIcon = [OACommonInteger withKey:locationIconKey defValue:LOCATION_ICON_DEFAULT];
-        [_locationIcon setModeDefaultValue:@(LOCATION_ICON_CAR) mode:OAApplicationMode.CAR];
-        [_locationIcon setModeDefaultValue:@(LOCATION_ICON_BICYCLE) mode:OAApplicationMode.BICYCLE];
-        [_locationIcon setModeDefaultValue:@(LOCATION_ICON_DEFAULT) mode:OAApplicationMode.BOAT];
-        [_locationIcon setModeDefaultValue:@(LOCATION_ICON_CAR) mode:OAApplicationMode.AIRCRAFT];
-        [_locationIcon setModeDefaultValue:@(LOCATION_ICON_BICYCLE) mode:OAApplicationMode.SKI];
-        [_locationIcon setModeDefaultValue:@(LOCATION_ICON_BICYCLE) mode:OAApplicationMode.HORSE];
+        _locationIcon = [OACommonString withKey:locationIconKey defValue:LOCATION_ICON_DEFAULT];
```
[the remark this answers is not in the document]
Comment (nnngrach) [reply Sources/Helpers/OAAppSettings.m:4000]: Yes. It translate old saved vars from Int to String:
Screenshot 2024-07-09 at 17 57 58

File: Sources/Helpers/Model3dHelper/Model3dHelper.swift [hunk 22]
```diff
@@ -0,0 +1,119 @@
...
+    private var modelsCache = [String: OAModel3dWrapper]()
+    private var modelsInProgress = Set<String>()
+    private var failedModels = Set<String>()
+    
+    private override init() {
+        app = OsmAndApp.swiftInstance()
+        settings = OAAppSettings.sharedManager()
+    }
+    
+    func getModel(modelName: String, callbackOnLoad: callbackWithModel3d?) -> OAModel3dWrapper? {
```
[the remark this answers is not in the document]
Comment (nnngrach) [reply Sources/Helpers/Model3dHelper/Model3dHelper.swift]: synced that commit https://github.com/osmandapp/OsmAnd-iOS/pull/3830/commits/6b040a61da0f6e83785e9454d4ee42b55a9ce3fd

File: Sources/Helpers/Model3dHelper/Model3dHelper.swift [hunk 23]
```diff
@@ -8,61 +8,95 @@
...
+
+    func loadAllPluginModels(callback: OAModel3dCallback?) {
+        let modelFoldersNames = Model3dHelper.listModels()
+        if !modelFoldersNames.isEmpty {
+            var loadingsCount = modelFoldersNames.count
             
-        } else {
-            loadModelImpl(modelName: modelName, callback: callback)
+            for modelName in modelFoldersNames {
+                getModel(modelName: modelName, callback: OAModel3dCallback { [weak self] model in
```
Comment (tigrim) [inline Sources/Helpers/Model3dHelper/Model3dHelper.swift:49]: remove wek self

File: Sources/Helpers/Model3dHelper/Model3dHelper.swift [hunk 24]
```diff
@@ -8,61 +8,95 @@
...
+        DispatchQueue.main.async { [weak self] in
+            self?.loadModelImpl(modelName: modelName, callback: callback)
+        }
+    }
+    
+    private func processCallback(modelName: String, model: OAModel3dWrapper?, callback: OAModel3dCallback?) {
+        if let callback {
+            if let callbacks = pendingCallbacks[modelName] {
+                if let index = callbacks.firstIndex(of: callback) {
+                    self.pendingCallbacks[modelName]?.remove(at: index)
```
Comment (tigrim) [inline Sources/Helpers/Model3dHelper/Model3dHelper.swift:69]: remove self

File: Sources/Helpers/Model3dHelper/Model3dHelper.swift [hunk 25]
```diff
@@ -8,61 +8,95 @@
...
+                    if !callbacks.isEmpty {
+                        for pendingCallback in callbacks {
+                            pendingCallback.processResult(model)
+                        }
+                    }
+                    self.pendingCallbacks.removeValue(forKey: modelName)
```
Comment (tigrim) [inline Sources/Helpers/Model3dHelper/Model3dHelper.swift:75]: remove self

File: Sources/Helpers/Model3dHelper/Model3dHelper.swift [hunk 26]
```diff
@@ -8,61 +8,95 @@
...
         }
-        
-        let modelDirPath = app.documentsPath.appendingPathComponent(MODEL_3D_DIR).appendingPathComponent(modelName)
-        if !Model3dHelper.isModelExist(dir: modelDirPath) {
+        if failedModels.contains(modelName) {
+            processCallback(modelName: modelName, model: nil, callback: callback)
+            return
+        }
+        if modelsInProgress.contains(modelName) {
+            if let callback {
```
Comment (tigrim) [inline Sources/Helpers/Model3dHelper/Model3dHelper.swift:92]: use 1 if

**Comment by nnngrach** [comment]
https://github.com/osmandapp/OsmAnd-iOS/assets/35684515/1b135774-2a6f-4b52-a2dc-32bbe20be70f

**Comment by nnngrach** [comment]
Partically synced this pr https://github.com/osmandapp/OsmAnd/pull/20229

Fixed import/export icon and color profile settings

export android -> ios
Screenshot 2024-07-18 at 22 52 59

export ios -> android
Screenshot 2024-07-18 at 22 54 06

[exported profiles.zip](https://github.com/user-attachments/files/16287137/exported.profiles.zip)

**Pull request merged**